Crowdin API v1 to v2 & Message extraction fixes

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

Updates the Crowdin script significantly to:
- Implement the v2 API for required commands (Upload sources, Download translations, Upload/Download Glossaries, Pre-translate, Build Translations)
- Adds some helpful commands to fetch information from the API as needed (list project, glossaries, files for branch)
- Adds a command to generate an HTML Screenshot report for cross checking screenshots against strings
- Adds a command to transfer screenshots that already exist for one branch on crowdin to another branch

Fixes bugs in message extraction:
- Aliased imports failing in studio
- Messages using a backticks but without any interpolation were being skipped because thats a `Literal` node and not a `TemplateLiteral` node...

**Does not port over from old script:**
- Upload translations. @radinamatic noted this was only really used for hacky testy purposes. Adding it if/when needed should be okay
- Stats. No comparable API endpoint in v2.

### API Module Approach

We use the  [Crowdin API Python Client](https://github.com/crowdin/crowdin-api-client-python) for API requests.

To handle Python dependencies, there is a utility that will run pip in a subprocess to install the dependency if it is not available.

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

Fixes #9678 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

@rtibbles let's find a time to walk through together - feel free to invite a +1 :)

nucleogenesis committed:
Migrate from v1 to v2 crowdin API.
Add utility for auto pip installing dependencies.

nucleogenesis committed:
i18n fe message extraction no longer dies, provides clearer error message for args, verbose option for info logs

nucleogenesis committed:
astUtils stringFromAnyLiteral takes a node, must be used to extract values

nucleogenesis committed:
fix bug where old extracted CSVs weren't able to be deleted

nucleogenesis committed:
clarify messaging aroudn searchPath, fix alias import following issues

nucleogenesis committed:
fix extraction from ObjectProperty

nucleogenesis commented:
@rtibbles successfully have tested this on Kolibri as well as Studio should be ready for final review

rtibbles commented on docs/i18n.rst:
```diff
@@ -258,7 +257,7 @@ This command will extract front- and backend strings and upload them to Crowdin,
 .. code-block:: bash
 
-  make i18n-upload branch=[release-branch-name]
+  make i18n upload-sources --branch=[release-branch-name]
```
I don't see any changes to the Makefile in this PR, so not sure how this command name has been updated, and even if it had, don't think it would have a space in it.

rtibbles commented on packages/kolibri-tools/lib/i18n/astUtils.js:
```diff
@@ -537,7 +539,8 @@ function getAstFromFile(filePath) {
 
   // Finally! Do the extraction
-  return parseAST(scriptContent);
+  let x = parseAST(scriptContent);
```
This looks like a change for debugging that didn't get fully reverted?

rtibbles commented on packages/kolibri-tools/lib/i18n/crowdin/cli.py:
```diff
@@ -216,5 +203,27 @@ def dispatch_command():
-    PARSED_ARGS = vars(parser.parse_args())
+
+    # Load the config file, update it with the parsed args (passed args overload config)
+    config = read_config_file("../../")
```
We should read from `os.cwd()` for the config file, otherwise this is very sensitive to where the package is installed for actually discovering the config (noting that if no `path` argument is passed, this is what `read_config_file` does by default anyway).

nucleogenesis replied:
The issue I'm having here I think is that the way the package is configured it's going to be run from wherever it is in the system because it isn't a PyPi package -- so the i18n Make command I had was like `cd packages/kolibri-tools && python -m lib.i18n.crowdin.cli ...` 

Maybe I'll just need to add a `config_path` option and put it on the user (us) to tell us where it is. I'll think on how to document this a bit better if that's how we do it

nucleogenesis replied:
The reasoning was that `node_modules/kolibri-tools/../../<Studio|KDP|etc>` and `packages/kolibri-tools/../../<Kolibri>` make this work out well enough but obviously is not ideal to rely on directory structure for a tool to work

rtibbles replied:
The other option would be to add an entry point in the package.json that invokes the python code, so that it can be run without the `cd`ing?

I am leaning slightly towards the idea of just publishing this as a little sidecar Python library, tbh, might make it easier.

rtibbles commented on packages/kolibri-tools/lib/i18n/crowdin/config.py:
```diff
@@ -144,6 +146,7 @@ def get_config():
                         --project-id, -p
                         --plugins-path [required only if --plugins-list is not given]
                         --plugins-list [required only if --plugins-path is not given]
+                        --locale-paths [required only if neighter of the above is provded]
```
```suggestion
                        --locale-paths [required only if neither of the above is provided]
```

rtibbles commented on packages/kolibri-tools/lib/i18n/crowdin/utils.py:
```diff
@@ -146,9 +146,9 @@ def json_dump_formatted(data, file_path, file_name):
+def read_config_file(path=None):
     output = {}
-    config_file = os.path.join(os.getcwd(), "setup.cfg")
+    config_file = os.path.join((path or os.getcwd()), "setup.cfg")
```
As I noted above, I don't think this change is necessary, and is likely to make the behaviour unstable.

rtibbles commented on packages/kolibri-tools/lib/i18n/crowdin/utils.py:
```diff
@@ -255,55 +255,58 @@ def _get_plugins_from_list(plugins_list):
-        plugins = _get_plugins_from_locale_paths(locale_paths)
-    return plugins
+        paths = _get_locale_paths_from_paths(locale_paths)
+    locale_data_folders = [path["locale_data_folder"] for path in paths]
```
Can also use a set comprehension here:
```suggestion
    locale_data_folders = {path["locale_data_folder"] for path in paths}
```

rtibbles commented on setup.cfg:
```diff
@@ -32,4 +32,8 @@ exclude_lines =
 locale_data_folder = kolibri/locale
+#
+# From the kolibri-tools package
+plugins_path = ../../build_tools/build_plugins.txt
```
I don't know what this file path is relative to, but it's not relative to the `setup.cfg`.

For clarity, and consistency with the `locale_data_folder` setting, we should make this `build_tools/build_plugins.txt` and handle that appropriately by resolving the path relative to the `setup.cfg`.

Do we need the locale_data_folder setting here any more, if we are deriving it from the build_plugins.txt?

rtibbles reviewed (dismissed):
Reading of the configuration file looks suspect to me, and I'd need quite some convincing that it's not.

Changes to documentation seem to contain an error unless I'm very much mistaken.

Rest is less blocking, but I'd prefer we smoothed out the remaining rough edges.

rtibbles committed:
Fix use/import of csvparse.

rtibbles committed:
Monkey patch node module resolution with any aliased paths.

rtibbles committed:
Add verbose handling to all function handlers.

rtibbles committed:
Add verbose to all handlers.
Unify options setting and pathInfo handling to add webpack aliases when needed.

rtibbles committed:
Add screenshot report functionality.

rtibbles committed:
Add command to transfer screenshots across branches.

rtibbles committed:
Add code comments to improve readability.

pcenov commented:
While routinely testing the build assets for this one I noticed that both exercises and html resources cannot be viewed. Just flagging it.

HTML5:

https://user-images.githubusercontent.com/79847249/215529161-a0e88738-b221-416a-beea-cf0c6192450f.mp4

EXERCISES:

https://user-images.githubusercontent.com/79847249/215529190-28ecdf59-db72-4b3a-a6df-2c72ab2f395c.mp4

radinamatic commented:
Testing the latest built DEB asset from the `develop` branch on Ubuntu, I can confirm both of the issues, so they are not caused by this PR. @pcenov could you please open a separate issue for these? Thank you!

<img width="1280" alt="2023-01-30_22-24-04" src="https://user-images.githubusercontent.com/1457929/215599090-516943b8-44f4-4f2c-9bdf-ed08e1dcad99.png">

One interesting thing: HTML resources do seem to fail to open, but after some time passes, completion is achieved, and learner can see the modal.

nucleogenesis commented on packages/kolibri-tools/lib/i18n/astUtils.js:
```diff
@@ -539,8 +510,7 @@ function getAstFromFile(filePath) {
   // Finally! Do the extraction
-  let x = parseAST(scriptContent);
-  return x;
+  return parseAST(scriptContent);
```
Pretty sure I assigned it so I could log it before returning but when I removed the log didn't really consider the pointlessness of `x` :smile:

nucleogenesis commented on packages/kolibri-tools/lib/cli.js:
```diff
@@ -448,359 +452,188 @@ program
+}
+
+// I18N Message Handling
+_addPathOptions(program.command('i18n-extract-messages')).action(function(options) {
+  const pathInfo = _generatePathInfo(options);
+  if (!pathInfo) {
+    program.command('i18n-extract-messages').help();
+  }
+  const extractMessages = require('./i18n/ExtractMessages');
+  extractMessages(pathInfo, options.ignore, options.localeDataFolder, options.verbose);
+});
+
+_addPathOptions(program.command('i18n-transfer-context')).action(function(options) {
+  const pathInfo = _generatePathInfo(options);
+  if (!pathInfo) {
+    program.command('i18n-transfer-context').help();
+  }
+  const syncContext = require('./i18n/SyncContext');
+  syncContext(pathInfo, options.ignore, options.localeDataFolder, options.verbose);
+});
```
Love this - much cleaner approach

nucleogenesis commented on packages/kolibri-tools/lib/i18n/crowdin.py:
```diff
@@ -90,8 +96,10 @@ def verify_branch(project_id, branch_name):
-        list_branches_response = crowdin_client.source_files.list_project_branches(
-            project_id
+        list_branches_response = (
+            crowdin_client.source_files.with_fetch_all().list_project_branches(
```
What is the outcome of not using `with_fetch_all` here? Would the response be paginated by default if you don't call this intermediary method?

nucleogenesis replied:
Mostly wondering if we should update any other crowdin API calls to use this.

rtibbles replied:
Yes, most of the calls are paginated by default - so any time we are listing something, we should probably be using this. I think I have updated most listing calls to use this in this PR.

nucleogenesis reviewed (commented):
Thanks @rtibbles for your work cleaning this up! 

I haven't tested it yet but wonder if maybe we should schedule a session w/ @radinamatic to walk through it and show her how it all works?

radinamatic commented:
> maybe we should schedule a session w/ @radinamatic to walk through

That would be most appreciated 🙏🏽

radinamatic commented:
~@nucleogenesis~ *rtibbles is on it*
 I bumped into `Error: Invalid value for "BRANCH": Branch is not valid` error.
I've put `apiv2` branch name just to test and delete later. Branch names should be able to contain both letters and numbers, so far they did, at least, correct?

```
$ make i18n-upload branch=apiv2
yarn run makemessages
yarn run v1.17.3
$ kolibri-tools i18n-extract-messages --pluginFile ./build_tools/build_plugins.txt
Gathering relevant modules from ['kolibri.core', 'kolibri.plugins.*']
Writing webpack_json output to /tmp/202312-10701-y0z9eh.282y.json
INFO: Writing webpack_json output to /tmp/202312-10701-y0z9eh.282y.json
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/core/**/*.@(vue|js)
Kolibri: Processing  227  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/core/**/*.@(vue|js)
Kolibri: Processing  227  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/user_profile/**/*.@(vue|js)
Kolibri: Processing  39  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/user_profile/**/*.@(vue|js)
Kolibri: Processing  39  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/media_player/**/*.@(vue|js)
Kolibri: Processing  32  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/facility/**/*.@(vue|js)
Kolibri: Processing  71  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/facility/**/*.@(vue|js)
Kolibri: Processing  71  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/epub_viewer/**/*.@(vue|js)
Kolibri: Processing  32  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/user_auth/**/*.@(vue|js)
Kolibri: Processing  30  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/user_auth/**/*.@(vue|js)
Kolibri: Processing  30  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/coach/**/*.@(vue|js)
Kolibri: Processing  243  files...
Kolibri: Skipping /home/didivm/LE/kolibri/kolibri/plugins/coach/assets/src/views/common/LearnerExerciseReport.vue because $trs property was given an empty object
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/coach/**/*.@(vue|js)
Kolibri: Processing  243  files...
Kolibri: Skipping /home/didivm/LE/kolibri/kolibri/plugins/coach/assets/src/views/common/LearnerExerciseReport.vue because $trs property was given an empty object
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/setup_wizard/**/*.@(vue|js)
Kolibri: Processing  51  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/html5_viewer/**/*.@(vue|js)
Kolibri: Processing  4  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/slideshow_viewer/**/*.@(vue|js)
Kolibri: Processing  4  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/demo_server/**/*.@(vue|js)
Kolibri: Processing  3  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/perseus_viewer/**/*.@(vue|js)
Kolibri: Processing  39  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/learn/**/*.@(vue|js)
Kolibri: Processing  170  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/learn/**/*.@(vue|js)
Kolibri: Processing  170  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/pdf_viewer/**/*.@(vue|js)
Kolibri: Processing  14  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/device/**/*.@(vue|js)
Kolibri: Processing  123  files...
Kolibri: Getting files from glob: /home/didivm/LE/kolibri/kolibri/plugins/device/**/*.@(vue|js)
Kolibri: Processing  123  files...
Kolibri: Removing existing messages files from /home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES
Kolibri: Successfully cleared path for CSVs by removing: /home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.core.default_frontend-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.core.frontend_head_assets-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.coach.app-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.coach.side_nav-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.demo_server.main-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.device.app-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.device.side_nav-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.epub_viewer.main-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.facility.app-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.facility.side_nav-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.html5_viewer.main-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.learn.app-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.learn.side_nav-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.media_player.main-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.pdf_viewer.main-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.perseus_viewer.main-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.setup_wizard.app-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.slideshow_viewer.main-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.user_auth.app-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.user_auth.login_side_nav-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.user_profile.app-messages.csv
/home/didivm/LE/kolibri/kolibri/locale/en/LC_MESSAGES/kolibri.plugins.user_profile.user_profile_side_nav-messages.csv
Kolibri: Successfully extracted 3574 messages!
Kolibri: Messages successfully written to CSV files.
Done in 37.72s.
cd kolibri && python -m kolibri manage makemessages -- -l en --ignore 'node_modules/*' --ignore 'kolibri/dist/*'
INFO     2023-02-02 11:25:47,404 Option DEBUG in section [Server] being overridden by environment variable KOLIBRI_DEBUG
INFO     2023-02-02 11:25:47,405 Option DEBUG_LOG_DATABASE in section [Server] being overridden by environment variable KOLIBRI_DEBUG_LOG_DATABASE
INFO     2023-02-02 11:25:47,405 Option RUN_MODE in section [Deployment] being overridden by environment variable KOLIBRI_RUN_MODE
INFO     2023-02-02 11:25:47,734 Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO     2023-02-02 11:25:47,998 Invoking command makemessages -l en --ignore node_modules/* --ignore kolibri/dist/*
processing locale en
python packages/kolibri-tools/lib/i18n/crowdin.py upload-sources apiv2
INFO: Project 'Kolibri - Learning Platform' found on Crowdin with ID '201933'
Branch apiv2 was not found. Would you like to create that branch? [y/N]: y
Usage: crowdin.py upload-sources [OPTIONS] BRANCH

Error: Invalid value for "BRANCH": Branch is not valid
Makefile:231: recipe for target 'i18n-upload' failed
make: *** [i18n-upload] Error 2
```

rtibbles committed:
Put confirm conditional in correct order.

rtibbles commented:
It seems I had 'yes' meaning 'no' and 'no' meaning 'yes' - which is liable to confuse anyone!

rtibbles committed:
Remove erroneous data keying.
Add with_fetch_all to most list endpoint requests.

rtibbles committed:
Add import options for new files as well as updated ones.

rtibbles committed:
Fix regression caused by change in pathInfo passed to extract message fn.

rtibbles committed:
Add missing scheme for CSV uploads.

rtibbles committed:
Fix bad positions for screenshots. Don't duplicate tags.

radinamatic reviewed (approved):
Several rounds of testing of the new commands and fixes in between: uploading to Crowdin, adding new translation and string context there, downloading and updating the codebase, **and** the new addition - transferring the screenshots from one version/folder to another on Crowdin, are all working beautifully!!! Thank you @nucleogenesis & @rtibbles 🏅 💯 :shipit:
